File: backend/models/match.go
package models

import (
	"time"
)

type MatchResult string

const (
	WhiteWin MatchResult = "white_win"
	BlackWin MatchResult = "black_win"
	Draw     MatchResult = "draw"
)

type Match struct {
	ID                uint        `gorm:"primaryKey" json:"id"`
	WhitePlayerID     uint        `gorm:"not null" json:"white_player_id"`
	BlackPlayerID     uint        `gorm:"not null" json:"black_player_id"`
	Result            MatchResult `gorm:"not null" json:"result"`
	PlayedAt          time.Time   `gorm:"not null" json:"played_at"`
	WhiteRatingBefore float64     `json:"white_rating_before"`
	BlackRatingBefore float64     `json:"black_rating_before"`
	WhiteRatingAfter  float64     `json:"white_rating_after"`
	BlackRatingAfter  float64     `json:"black_rating_after"`
	ChesscomGameID    *string     `gorm:"uniqueIndex" json:"chesscom_game_id"`
	PGN               *string     `gorm:"type:text" json:"pgn"`
	CreatedAt         time.Time   `json:"created_at"`

	// Relations
	WhitePlayer User `gorm:"foreignKey:WhitePlayerID" json:"white_player,omitempty"`
	BlackPlayer User `gorm:"foreignKey:BlackPlayerID" json:"black_player,omitempty"`
}

type CreateMatchRequest struct {
	WhitePlayerID  uint        `json:"white_player_id" binding:"required"`
	BlackPlayerID  uint        `json:"black_player_id" binding:"required"`
	Result         MatchResult `json:"result" binding:"required,oneof=white_win black_win draw"`
	PlayedAt       time.Time   `json:"played_at" binding:"required"`
	ChesscomGameID *string     `json:"chesscom_game_id"`
	PGN            *string     `json:"pgn"`
}

type BulkCreateMatchRequest struct {
	Matches []CreateMatchRequest `json:"matches" binding:"required,dive"`
}

File: frontend/src/types/match.ts
import { User } from './user';

export type MatchResult = 'white_win' | 'black_win' | 'draw';

export interface Match {
  id: number;
  white_player_id: number;
  black_player_id: number;
  result: MatchResult;
  played_at: string;
  white_rating_before: number;
  black_rating_before: number;
  white_rating_after: number;
  black_rating_after: number;
  chesscom_game_id?: string;
  white_player?: User;
  black_player?: User;
  created_at: string;
}

export interface CreateMatchRequest {
  white_player_id: number;
  black_player_id: number;
  result: MatchResult;
  played_at: string;
  chesscom_game_id?: string;
}

export interface BulkCreateMatchRequest {
  matches: CreateMatchRequest[];
}

// Chess.com types
export interface ChesscomPlayer {
  username: string;
  rating: number;
  result: string;
}

export interface ChesscomGame {
  url: string;
  game_id: string;
  pgn: string;
  end_time: number;
  time_class: string;
  rated: boolean;
  white: ChesscomPlayer;
  black: ChesscomPlayer;
}

export interface ChesscomGamesResponse {
  games: ChesscomGame[];
  already_imported: string[];
  user1: {
    id: number;
    name: string;
    chesscom_username: string;
  };
  user2: {
    id: number;
    name: string;
    chesscom_username: string;
  };
}

export interface ChesscomSyncResponse {
  year: number;
  month: number;
  recalculate: boolean;
  recalculate_done: boolean;
  users_considered: number;
  total_fetched: number;
  total_between_players: number;
  already_imported: number;
  created_count: number;
}

export interface ChesscomUserProfile {
  player_id: number;
  username: string;
  name?: string;
  status: string;
  country?: string;
  joined: number;
  last_online: number;
}
